optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsGroupingBiConstraintStream.java

```diff
@@ -0,0 +1,104 @@
+    private final Function<A, NewA> groupKeyMapping;
+    private final UniConstraintCollector<A, ResultContainer_, NewB> collector;
+    private final AtomicInteger ruleId = new AtomicInteger(-1);
+
+    public DroolsGroupingBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractUniConstraintStream<Solution_, A> parent, Function<A, NewA> groupKeyMapping,
+            UniConstraintCollector<A, ResultContainer_, NewB> collector) {
+        super(constraintFactory);
+        this.parent = parent;
+        this.groupKeyMapping = groupKeyMapping;
+        this.collector = collector;
+    }
+
+    @Override
+    public Optional<Rule> buildRule(DroolsConstraint<Solution_> constraint,
+            Global<? extends AbstractScoreHolder<?>> scoreHolderGlobal) {
+        Object createdRuleId = createRuleIdIfAbsent(constraint.getConstraintFactory());
+        String ruleName = "Helper rule #" + createdRuleId + " (GroupBy remapping+collecting)";
+        Rule rule = PatternDSL.rule(constraint.getConstraintPackage(), ruleName)
+                .build(parent.createCondition().completeWithLogicalInsert(createdRuleId, groupKeyMapping, collector)
+                        .toArray(new RuleItemBuilder<?>[0]));
+        return Optional.of(rule);
+    }
+
+    @Override
+    public DroolsBiCondition<NewA, NewB> createCondition() {
+        return new DroolsBiCondition<>(declarationOf(DroolsLogicalTuple.class), var -> {
+            Object createdRuleId = createRuleIdIfAbsent(getConstraintFactory());
+            return pattern(var).expr(logicalTuple ->
+                    Objects.equals(logicalTuple.getRuleId(), createdRuleId));
+        });
+    }
+
+    /**
+     * Exists in order to be able to bind the condition in the new rule to the {@link DroolsLogicalTuple}s generated by
+     * the old rule, while only generating the rule ID on-demand to maintain the requirement of rule IDs being in an
+     * unbroken sequence.
+     * The idea here is that rule creating ({@link DroolsAbstractConstraintStream#buildRule(DroolsConstraint, Global)})
+     * would call this first, establishing the rule ID with which to insert all the {@link DroolsLogicalTuple}s.
+     * The later creation of the condition during {@link DroolsAbstractBiConstraintStream#createCondition()} would
+     * just retrieve the value that already exists and query for all the {@link DroolsLogicalTuple}s having this value.
+     * Even if the order of operations were switched (condition created before rule built), the ID would still be
+     * correctly shared though.
+     * @param constraintFactory never null, used to generate rule IDs
+     * @return unique id for the rule
+     */
+    private int createRuleIdIfAbsent(DroolsConstraintFactory<Solution_> constraintFactory) {
+        return ruleId.updateAndGet(currentRuleId -> (currentRuleId < 0) ?
+                constraintFactory.getRuleIdAndIncrement() :
+                currentRuleId);
```

Comment:
Took me a while to understand this.
See previous comment, it could be made:

```
if (ruleId < 0) {
   ruleId = constraintFactory.getRuleIdAndIncrement();
}
```

Btw, `constraintFactory.getRuleIdAndIncrement()` might be worth renaming to `constraintFactory.getNextRuleId()`. The incrementing is an impl detail.